File: book-import/src/main/java/com/luanvv/microservices/bookimport/BookService.java
package com.luanvv.microservices.bookimport;

import lombok.AllArgsConstructor;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@AllArgsConstructor
public class BookService {

    private final JdbcTemplate jdbcTemplate;

    @Transactional
    public int save(Book book) {
        List<Integer> ids = jdbcTemplate.query("SELECT id FROM book WHERE title = ? LIMIT 1", (rs, rnum) -> rs.getInt(1),
                book.getTitle());
        if (ids.isEmpty()) {
            return jdbcTemplate.update(
                    "INSERT INTO book (title, author, genre, height, publisher) VALUES(?, ?, ?, ?, ?)",
                    book.getTitle(), book.getAuthor(), book.getGenre(), book.getHeight(), book.getPublisher());
        } else {
            return jdbcTemplate.update("UPDATE book SET author=?, genre=?, height=?, publisher=? WHERE id=?",
                    book.getAuthor(), book.getGenre(), book.getHeight(), book.getPublisher(), ids.get(0));
        }

    }

}

File: product-import/src/main/kotlin/com/luanvv/microservices/bookstore/product_import/services/BookService.kt
package com.luanvv.microservices.bookstore.product_import.services

import com.luanvv.microservices.bookstore.product_import.entities.Book
import com.luanvv.microservices.bookstore.product_import.repositories.BookRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class BookService(val repository: BookRepository) {

    @Transactional
    fun save(newBook: Book) {
        val existBook = repository.findByIsbn10(newBook.isbn10)
        val willSaveBook = existBook?.update(newBook) ?: newBook
        repository.save(willSaveBook)
    }

    private fun Book.update(book: Book): Book {
        title = book.title
        author = book.author
        genre = book.genre
        height = book.height
        publisher = book.publisher
        isbn10 = book.isbn10
        status = book.status
        return this
    }
}